enhancements/proposals/cleanup_after_scaling.md

```diff
@@ -0,0 +1,140 @@
+ScyllaDB plans to automate cleanup and cover all the cases, until then we won’t cover these cases.
+
+## Design Details
+
+Scylla-operator will be extended with a controller responsible for running a cleanup procedure when needed.
+
+### Cleanup trigger
+
+The root cause of the cleanup requirement is that during cluster changes tokens are moved between nodes.
+There are two alternatives of detecting when this happened:
+
+1. ScyllaCluster topology changes - new rack is added/removed, number of members changes, or additional CPUs are
+   added/removed.
+2. Check tokens in Scylla API, detect when it changed.
+
+The first one is edge driven, there are multiple cases when tokens might move or not, for example if cpu resource change
+won’t be big enough to allocate additional core the cleanup is not needed. For hybrid clusters or multi datacenter
+clusters, topology change in the non-managed part should cause the cleanup of nodes in the managed part which is
+impossible to detect. It’s also harder to implement since StatefulSets, which changes need to be detected, doesn't allow
+for per node distinction, the state needs to be carried to other resources (for example member Service) which brings
+unnecessary coupling.
+
+The second one is checking the actual root cause, and seems more bullet proof.
+
+### Implementation
+
+Two new annotations in member services will be reconciled. Controller in the sidecar will
+reconcile `internal.scylla-operator.scylladb.com/current-token-ring-hash`
+and Operator will reconcile `internal.scylla-operator.scylladb.com/last-cleaned-up-token-ring-hash`.
+
+The `current` one will always reflect the hashed output of Scylla API `/storage_service/tokens` endpoint from a given
+node. The `last` will be either equal to `current` if it’s not yet present, or will be updated to current only when
+cleanup
+Job for a given node is `Completed`.
+
+The output of tokens endpoint is an json sorted array of strings representing numbers, the annotation will keep SHA512
+hash of it as a value.
+
+The cleanup call is synchronous, client only triggers the operation, and all the resource intensive operations are done
+by Scylla. The duration of the operation is tied to the size of the entire dataset and disk performance. Cleanup on node
+that don't require it, also impacts the disk IO.
+To prevent the controller sync loop from being too long, the cleanup controller in Operator will be responsible for
+managing cleanup Jobs. The preconditions for Job for given node are:
+
+* `current` and `last` token ring hash annotations values are different
+* All previous nodes have their `current` and `last` token ring hash annotations values equal - the order is rack order
+  in ScyllaCluster.spec and Pod ordinal - to prevent two cleanup Jobs running concurrently.
+* StatefulSet controller is not progressing - topology changes are done.
+* All nodes in ScyllaCluster are UN - every rack status readyMembers matches the number of members in spec.
```

Comment:
how will the controller check this without a client? through readiness state (which is already in SC conditions)?

Reply:
Added mention about how it's checked

Reply:
progressing==False doesn't mean all the nodes are UN, that's why I've suggested the other conditions as well.
Technically, this should also check generation for the status it sees.

There is also the question of whether we'd do a live call in case the informers are stale.